### ISSUE TITLE
oasis installs fails (lib/userconf and lib/oasis already exist)

### ISSUE THREAD

**Issue by tizoc** [ocaml/opam#2366]
This started happening after an upgrade. It seems opam didn't clean up the old installs, and thats causing problems when trying to install the new version:

```
#=== ERROR while installing oasis.0.4.5 =======================================#
# opam-version 1.2.2
# os           darwin
# command      ocaml setup.ml -install
# path         /Users/bruno/.opam/system/build/oasis.0.4.5
# compiler     system (4.02.3)
# exit-code    1
# env-file     /Users/bruno/.opam/system/build/oasis.0.4.5/oasis-3595-c61e7b.env
# stdout-file  /Users/bruno/.opam/system/build/oasis.0.4.5/oasis-3595-c61e7b.out
# stderr-file  /Users/bruno/.opam/system/build/oasis.0.4.5/oasis-3595-c61e7b.err
### stderr ###
# File "src/oasis/OASISString.ml", line 148, characters 8-26:
# Warning 3: deprecated: String.set
# Use Bytes.set instead.
# File "src/oasis/OASISFindlib.ml", line 262, characters 6-24:
# Warning 3: deprecated: Lazy.lazy_from_fun
# Use Lazy.from_fun instead.
# ocamlfind: Package userconf is already installed
#  - (file /Users/bruno/.opam/system/lib/userconf/META already exists)
# E: Failure("Command ''/Users/bruno/.opam/system/bin/ocamlfind' install userconf src/ext/userconf/src/META _build/src/ext/userconf/src/UserConf.cmx _build/src/ext/userconf/src/UserConf.cmi _build/src/ext/userconf/src/userconf.cmxs _build/src/ext/userconf/src/userconf.a _build/src/ext/userconf/src/userconf.cmxa _build/src/ext/userconf/src/userconf.cma src/ext/userconf/src/UserConf.ml' terminated with error code 2")
```

I moved those directories (`lib/oasis`, `lib/plugin-loader` and `lib/userconf`) somewhere else, and after doing so the install succeeded.

**Comment by Chris00** [comment]
May you post the result of `opam show oasis --raw`?

**Comment by tizoc** [comment]
Here it is, but keep in mind that I deleted the old directories and reinstalled oasis, so the output I get now may not be what you wanted?

```
opam-version: "1.2"
name: "oasis"
version: "0.4.5"
maintainer: "Sylvain Le Gall <gildor@ocamlcore.org>"
authors: "Sylvain Le Gall"
homepage: "http://oasis.forge.ocamlcore.org/"
bug-reports: "http://oasis.forge.ocamlcore.org/"
license: "LGPL-2.1 with OCaml linking exception"
dev-repo: "git://github.com/gildor478/oasis.git"
build: [
  [
    "ocaml"
    "setup.ml"
    "-configure"
    "--prefix"
    prefix
    "--%{gettext:enable}%-gettext"
  ]
  ["ocaml" "setup.ml" "-build"]
]
install: ["ocaml" "setup.ml" "-install"]
build-test: [
  [
    "ocaml"
    "setup.ml"
    "-configure"
    "--enable-tests"
    "--%{gettext:enable}%-gettext"
  ]
  ["ocaml" "setup.ml" "-build"]
  ["ocaml" "setup.ml" "-test"]
]
build-doc: ["ocaml" "setup.ml" "-doc"]
remove: [
  "/bin/sh"
  "-c"
  "if [ -e '%{etc}%/oasis/_oasis_remove_.ml' ]; then ocaml '%{etc}%/oasis/_oasis_remove_.ml' '%{etc}%/oasis'; else ocamlfind remove plugin-loader; ocamlfind remove userconf; ocamlfind remove oasis; rm '%{bin}%/oasis'; fi"
]
depends: [
  "base-unix"
  "expect" {test & >= "0.0.4"}
  "fileutils" {test & >= "0.4.2"}
  "ocaml-data-notation" {>= "0.0.11"}
  "ocamlfind" {>= "1.3.1"}
  "ocamlify" {build}
  "ocamlmod" {build}
  "ounit" {test & >= "2.0.0"}
  "pcre" {test}
]
conflicts: [
  "oasis-mirage" {= "0.3.0a"}
  "oasis-mirage" {= "0.3.0"}
]
available: [ocaml-version >= "3.11.2"]
```

**Comment by Chris00** [comment]
This is the new version of the file so I suppose you cannot reproduce the problem.

I am a bit surprised that you encountered the problem because the rule

    remove: [
      "/bin/sh"
      "-c"
      "if [ -e '%{etc}%/oasis/_oasis_remove_.ml' ]; then ocaml '%{etc}%/oasis/_oasis_remove_.ml' '%{etc}%/oasis'; else ocamlfind remove plugin-loader; ocamlfind remove userconf; ocamlfind remove oasis; rm '%{bin}%/oasis'; fi"
    ]

is supposed to perform what you did by hand when %{etc}%/oasis/_oasis_remove_.ml is not present — which is the case for an old installation.  Or did you do an "opam update" after having the problem?

**Comment by tizoc** [comment]
I don't remember if I did an `opam update` after that, I know that the version of oasis didn't change, but the package metadata may have changed.

Someone on the #ocaml channel @ freenode was having the exact same problem as me, but this was a few days ago, it is unlikely that by now they have kept their install unchanged.

For that it is worth, I remember that the directories where still there under `lib/` but that the `oasis` binary was not present, so it is possible that the `ocamlfind remove` commands had failed (but the `rm` worked fine).

**Comment by Chris00** [comment]
You can try to remove `~/.opam/system/etc/oasis/_oasis_remove_.ml` (replace `system` with your switch) and try to remove oasis to see whether the ocamlfind directories are still present.

**Comment by AltGr** [comment]
I don't think this is a problem in oasis, on the flambda bench tests I have from time to time uncleanly removed packages that cause ocamlfind issues when reinstalling. It seems to happen very randomly, and I dont think it is a bug in opam. Note that simply retrying usually solves the problem (because when install fails, opam runs the uninstall instructions to be sure that no artifact remains).

My hunch is that ocamlfind doesn't uninstall properly when run in certain circumstances; maybe when it's in parallel. So 3746a1c4aa8eae50413e743482d2024ae0a1af7c _might_ fix this...

**Comment by tizoc** [comment]
@Chris00 I removed `_oasis_remove_.ml` and did an `opam uninstall oasis`, the first time I got this warning:

```
=-=- Gathering sources =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫 
[WARNING] The sources of the following couldn't be obtained, they may be uncleanly uninstalled:
            - oasis.0.4.5
```

But it was uninstalled cleanly anyway. I had to do an `opam update` to be able to install it again because the sources could not be found.

I installed it again, and repeated the process again, worked fine again, and this time without any warning.

This probably has nothing to do with the issue I had before, but mentioning it just in case.
